### PR TITLE
Feature delete tag

### LINKED ISSUE #40: Delete tag
Add tag deletion support to improve repository image management.

### PR DESCRIPTION
**Pull request by Vasilijez**
Closes #40. 

I’ve implemented `RegistryClient` as a single shared instance across the application’s lifecycle to improve efficiency and performance.

I adjusted the certificates to support seamless communication between the backend and distribution containers.

I have chosen to run Distribution GC before the container starts, as it can be dangerous if Distribution is not in [read-only mode](https://distribution.github.io/distribution/about/garbage-collection/#more-details-about-garbage-collection). I’ve also provided a variable to enable or disable this behavior, in case running it each time is too frequent.

I decided to remove the `base image` label, as it is very difficult to map it to the underlying Distribution concepts and mechanisms — even though it initially seemed like a good idea. I'll try to explain why this is the case.
A `base image` doesn't truly exist. What we have is a manifest (an image) that links to all the used blobs (layers). Layers are shared among manifests (images), and mounting layers is a common and frequent action. Some layers may become unused when an image is deleted, while others remain in use because they are still mounted elsewhere. Therefore, it's very hard to define a single "base image" for any given image. I can provide more examples if needed.

### CODE REVIEW

File: .github/workflows/run_tests.yml [hunk 1]
```diff
@@ -24,11 +24,11 @@ jobs:
...
-    - name: Unpack certificates
-      run: |
-        sudo apt-get update
-        sudo apt-get install -y unrar
-        unrar x distribution/certs/certs.rar distribution/certs/
+    # - name: Unpack certificates
+    #   run: |
+    #     sudo apt-get update
+    #     sudo apt-get install -y unrar
+    #     unrar x distribution/certs/certs.rar distribution/certs/
```
Comment (magley) [inline .github/workflows/run_tests.yml]: This is commented out, but will we ever need this in the future? `entrypoint.sh` extracts the zip anyway.

File: README.md [hunk 2]
```diff
@@ -71,14 +71,15 @@ In case you still have issues, create a file `/etc/docker/daemon.json` and write
...
+docker push localhost:5000/{image-name}[:{tag}]
 ```
+If no `:tag` is provided, latest is used by default.
 
 Pull an image from the registry:
 
 ```sh
-docker pull localhost:5000/{image-name}
+docker pull localhost:5000/{image-name}[:{tag}]
 ```
```
Comment (magley) [inline README.md:83]: Can we delete tags from the command line now? If so, you could add an example in the readme here.

File: client/src/components/RepoTags.tsx [hunk 3]
```diff
@@ -60,6 +63,23 @@ export const RepoTags: React.FC<{ isActive: boolean, repo: RepoExtDTO }> = (prop
...
+                    // setError((err.response?.data as any)["detail"]["message"]);
+                    const data = (err.response?.data ?? {}) as any;
+                    const msg = typeof data.detail === 'string' ? data.detail : data.detail?.message;
+                    addToast(msg, ToastType.error);
+                });
+        } else {
+            addToast('Tag name is missing and cannot be deleted.', ToastType.error);
+        }
+    };
+
```
Comment (magley) [inline client/src/components/RepoTags.tsx]: If deleting tags takes too long, we should temporarily disable the button and add a spinner.

I haven't tested this myself, so I don't know how much time it takes for a "delete tag" request to process, so this comment is just a suggestion - if you feel this is unnecessary then you can just leave it as is.

File: client/src/components/RepoTags.tsx [hunk 4]
```diff
@@ -104,29 +124,35 @@ export const RepoTags: React.FC<{ isActive: boolean, repo: RepoExtDTO }> = (prop
...
+                                    {/* ... */}
+                                    <Card.Text style={{ fontSize: '0.8rem' }}>
+                                    </Card.Text>
+                                </div>
+
+                                {/* Delete tag */}
+                                {/* We can reuse `can_update` as `can_delete_tag` since the access-control logic is identical. */}
+                                {props.repo?.can_update && tag.name &&
+                                    <Button variant="danger" className="me-2" onClick={() => handleDeleteTag(tag)} title="Delete tag">
```
Comment (magley) [inline client/src/components/RepoTags.tsx]: `can_update` and `can_delete_tag` are equivalent as a consequence of the project spec. I was thinking of creating a `can_delete` alias, just for readability. What I was hoping to do was something like this:

```ts
interface RepoExtDTO {
    ...
    can_update: bool
    can_delete => can_update
}
```

But typescript doesn't support that. The best we can do is use a getter, except that's only allowed in classes and not interfaces.

The only other option I can think of is to give the RepoTags component a `can_delete` prop. This is ugly IMO.

Well, another option is to create a `can_delete` field in the DTO and have the server populate that with a true/false. That's the "right" solution, but since we sort of know that update and delete require the same privileges, it feels redundant.

---

Ultimately, there's nothing you should change here, but I wanted to leave this comment because I feel like we should separate `update` and `delete`. Yes, it's true that they are equivalent on the _server_ side, but the _client_ side should not make such assumptions.

File: compose.yaml [hunk 5]
```diff
@@ -94,6 +97,10 @@ services:
     volumes:
       - "./distribution/config.yaml:/etc/docker/registry/config.yml"
       - "./distribution/certs:/mnt/local/certs"
+      - "./distribution/entrypoint.sh:/entrypoint.sh"
+    entrypoint: ["/bin/sh", "/entrypoint.sh"]
+    environment:
+      RUN_REGISTRY_GC: "true"
```
Comment (magley) [inline compose.yaml:103]: Is it safe to run the GC every time? How come this isn't the default behavior?

File: server/app/api/config/auth.py [hunk 6]
```diff
@@ -163,12 +163,15 @@ def pre_authorize(roles: List[str] | List[UserRole] | None, ignore_password_chan
...
             jwt_dep = kwargs.get('jwt')
             if not jwt_dep:
                 # TODO: Change error message in production.
                 raise Exception("Cannot perform authorization without a JWT. Your router endpoint function MUST have a parameter named `jwt: JWTDep`")
             validate_jwt_or_raise_exceptions(jwt_dep, roles, ignore_password_change_requirement)
-            return func(*args, **kwargs)
+            if inspect.iscoroutinefunction(func):
+                return await func(*args, **kwargs)
+            else:
+                return func(*args, **kwargs)
```
Comment (magley) [inline server/app/api/config/auth.py:175]: This is a cool trick. It's hard for me wrap my head around why this is allowed, to have an async wrapper of a maybe-async/maybe-sync function. Reminds me of the article [What Color is Your Function?](https://journal.stuffwithstuff.com/2015/02/01/what-color-is-your-function/).

File: server/app/api/registry/registry_controller.py [hunk 7]
```diff
@@ -25,17 +34,43 @@ def registry_endpoint(
...
+    user_id = get_id_from_jwt(jwt)
+    username = get_username_from_jwt(jwt)
+    repo = repo_service.find_by_id(dto.repo_id)
+    tag = tag_service.find_by_name_and_repo_id(dto.tag_name, dto.repo_id)
+
+    if not access_control_service.has_delete_tag_access(user_id, repo.id):
+        raise AccessDeniedException(f"User {user_id} cannot delete a tag {tag.name} of repository with identifier {repo.id}")
+
+    response = await registry_service.delete_tag(registry_client, username, repo, tag)   
+
```
Comment (magley) [inline server/app/api/registry/registry_controller.py:75]: Why do we check for access control here instead of inside `registry_service.delete_tag()`?

File: server/app/api/repo/repo_service.py [hunk 8]
```diff
@@ -129,6 +129,9 @@ def toggle_repo_star(self, user_id: int, repo_id: int) -> Tuple[Repository, bool
 
         repo = self.repo_repo.unstar_repo(repo, user)
         
+        if repo is None:
+            raise ConflictException("The repository star was removed (unstarred) by a different operation")
```
Comment (magley) [inline server/app/api/repo/repo_service.py:133]: This seems a bit too verbose. Furthermore, `repo_repo.unstar_repo()` will return `None` only when `user` didn't star `repo`, so the message is also a bit misleading (if the repo was never starred by the user, that doesn't mean that there was a different operation that unstarred earlier).

Does the message propagate back to the controller (and all the way to the client)? If that's the case, we should consider just telling the client that "there was an error".

This is me being nitpicky, it ultimately doesn't matter.

**Review by magley — CHANGES_REQUESTED** [review]
Clever use of internal_router and external_router. I also appreciate that you delete manifests from Distribution first, before touching the database.

Note: I didn't run this code on my computer. 

The suggestions I left are mostly comments, stylistic improvements. I see no logic errors, but part of that is due to me still not exactly knowing how tags, images, manifests etc. correlate and how they are stored by Distribution i.e. a lack of fundamentals.

File: .github/workflows/run_tests.yml (hunk 1 above)
Comment (Vasilijez) [reply .github/workflows/run_tests.yml]: Haha, you are right, I accidently left those lines in.

https://github.com/magley/mocker-hub/pull/43/commits/d98c81038ba228baf882f29f3fde018b1474d184

File: README.md (hunk 2 above)
Comment (Vasilijez) [reply README.md:83]: That would be great, but unfortunately, it's not possible.

File: client/src/components/RepoTags.tsx (hunk 3 above)
Comment (Vasilijez) [reply client/src/components/RepoTags.tsx]: Deleting a tag usually takes only a few milliseconds, since it mainly involves removing pointers to image manifests. However, just in case, I'll implement it. It seems like a good idea, and I was already debating whether or not to do it.

https://github.com/magley/mocker-hub/pull/43/commits/6d85e5add112978ab7bad4d9db5b1121ed542d99

File: client/src/components/RepoTags.tsx (hunk 4 above)
Comment (Vasilijez) [reply client/src/components/RepoTags.tsx]: I really like this comment, as it highlights the importance of separating client-side and server-side logic.

File: compose.yaml (hunk 5 above)
Comment (Vasilijez) [reply compose.yaml:103]: Haha, yes — it's completely safe to perform this on every restart. It will simply attempt to identify any blobs that are no longer referenced and remove them if found. That’s a great question for the Distribution developers. From what I understand, this is intentional. In most use cases, it’s better not to delete blobs immediately, since they are often mounted/shared across images. Also, deleting tags or repositories is relatively rare.

File: server/app/api/config/auth.py (hunk 6 above)
Comment (Vasilijez) [reply server/app/api/config/auth.py:175]: Haha, glad you found it interesting!

File: server/app/api/registry/registry_controller.py (hunk 7 above)
Comment (Vasilijez) [reply server/app/api/registry/registry_controller.py:75]: We already discussed this comment in private messages.

File: server/app/api/repo/repo_service.py (hunk 8 above)
Comment (Vasilijez) [reply server/app/api/repo/repo_service.py:133]: As I mentioned earlier, this test is a bit silly and unnecessarily complex.

The user had already starred the repository before calling `self.repo_repo.unstar_repo(repo, user)`, as indicated by the preceding if `repo_is_starred == False:` check. So this test is really about a concurrency edge case — specifically, when a `RepositoryStar` object is unexpectedly missing from the session.

It's an extremely rare case that depends on the transaction isolation level and would require a deeper investigation. I imagined a scenario where the user quickly sends two consecutive requests by clicking twice, which could potentially lead to this issue. 

Yes, the error will propagate back to the client. But if you still feel that the message isn't appropriate for the client, I'm happy to change it to something more generic like "There was an error."

File: README.md (hunk 2 above)
Comment (magley) [reply README.md:83]: What about [`docker rmi image:tag`](https://serverfault.com/a/703352)?

File: server/app/api/registry/registry_controller.py (hunk 7 above)
Comment (magley) [reply server/app/api/registry/registry_controller.py:75]: [Note: we came to the conclusion that it _should_ be in the service layer because it's business logic. `pre_authorize` being in the controller layer is a remnant of Spring Boot. Unfortunately, there are many other places where we broke this rule, and seeing as it's not a major issue, it's best to leave it alone and not create additional work]

File: README.md (hunk 2 above)
Comment (Vasilijez) [reply README.md:83]: That command does not remove images from a registry, it only works for local images.

Some references:
- [docker image rm](https://docs.docker.com/reference/cli/docker/image/rm/)
- [remove tagged Docker image from private registry](https://serverfault.com/questions/750094/remove-tagged-docker-image-from-private-registry)

Here’s what happens when we run the command on our remote registry:

image

Comment (magley) [reply README.md:83]: Alright then, I guess it makes sense not to be able to remove tags from the command line. You can't delete repositories from git cli either.

**Review by magley — APPROVED** [review]
